lib/web/device_trust.go

```diff
@@ -76,7 +81,29 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
 
 	// Always redirect back to the dashboard, regardless of outcome.
 	app.SetRedirectPageHeaders(w.Header(), "" /* nonce */)
-	http.Redirect(w, r, "/web", http.StatusSeeOther)
+
+	redirectTo := "/web"
+
+	if redirectParam != "" {
+		parsedURL, err := url.Parse(redirectParam)
+		if err == nil {
+			cleanPath := path.Clean(parsedURL.Path)
+			// helps in situations where there is no path such as https://example.com
+			if cleanPath == "." || cleanPath == "" {
+				cleanPath = "/"
+			} else if !strings.HasPrefix(cleanPath, "/") {
+				cleanPath = "/" + cleanPath
+			}
+			// Prepend "/web" only if it's not already present
+			if !strings.HasPrefix(cleanPath, "/web") {
+				redirectTo = path.Join("/web", cleanPath)
+			} else {
+				redirectTo = cleanPath
+			}
```

Comment:
Let's make sure we test all the corner cases of this block. It seems a bit tricky to get right.